aws-lc-rs/src/cipher/streaming.rs

```diff
@@ -0,0 +1,786 @@
+        if 1 != unsafe {
+            EVP_EncryptInit_ex(
+                cipher_ctx.as_mut_ptr(),
+                *cipher,
+                null_mut(),
+                key_bytes.as_ptr(),
+                iv.as_ptr(),
+            )
+        } {
+            return Err(Unspecified);
+        }
+
+        Ok(Self {
+            algorithm,
+            mode,
+            cipher_ctx,
+            context,
+        })
+    }
+
+    /// Updates the internal state of the key with the provided ciphertext `input`,
+    /// potentially writing bytes of ciphertext to `output`.
+    ///
+    /// The number of bytes written to `output` can be up to `input.len()`
+    /// plus the block length of the algorithm (e.g., [`Algorithm::block_len`]).
+    ///
+    /// # Errors
+    /// * Returns an error if the `output` buffer is smaller than the length of
+    ///   the `input` plus the algorithm's block length (e.g. [`Algorithm::block_len`]) minus one.
+    /// * May return an error if the length of `input` plus the algorithm's block length is larger than `i32::MAX`.
+    pub fn update<'a>(
+        &mut self,
+        input: &[u8],
+        output: &'a mut [u8],
+    ) -> Result<BufferUpdate<'a>, Unspecified> {
+        let min_outsize = input
+            .len()
+            .checked_add(self.algorithm().block_len())
+            .ok_or(Unspecified)?
+            - 1;
+        if output.len() < min_outsize {
+            return Err(Unspecified);
+        }
+        let mut outlen: i32 = 0;
+        let inlen: i32 = input.len().try_into()?;
+
+        if 1 != unsafe {
+            EVP_EncryptUpdate(
+                self.cipher_ctx.as_mut_ptr(),
+                output.as_mut_ptr(),
+                &mut outlen,
+                input.as_ptr(),
+                inlen,
+            )
+        } {
+            return Err(Unspecified);
+        }
+        let outlen: usize = outlen.try_into()?;
+        Ok(BufferUpdate::new(output, outlen))
+    }
+
+    /// Finishes the encryption operation, writing any remaining ciphertext to
+    /// `output`.
+    ///
+    /// The number of bytes written to `output` can be up to the block length of
+    /// [`Algorithm::block_len`].
+    ///
+    /// # Errors
+    /// * Returns an error if the `output` buffer is smaller than the algorithm's
+    ///   block length.
+    pub fn finish(
+        mut self,
+        output: &mut [u8],
+    ) -> Result<(DecryptionContext, BufferUpdate), Unspecified> {
+        if output.len() < self.algorithm().block_len() {
+            return Err(Unspecified);
+        }
+        let mut outlen: i32 = 0;
+
+        if 1 != indicator_check!(unsafe {
+            EVP_EncryptFinal_ex(
+                self.cipher_ctx.as_mut_ptr(),
+                output.as_mut_ptr(),
+                &mut outlen,
+            )
+        }) {
+            return Err(Unspecified);
+        }
+        let outlen: usize = outlen.try_into()?;
```

Comment:
Same note here on whether we should assert `outlen` being less then or equal to `output.len()`.